### PR TITLE
[KOGITO-4677] Fix code smell for explainability module

### PR DESCRIPTION
**Pull request by danielezonca**
https://issues.redhat.com/browse/KOGITO-4677

## Details
- Removed generic from `Value` because it was ~always used with `<?>` or raw type
- Fixed most (all?) code smell in `explainability` module
- Refactored counterfactual `*Entity` classes to reduce code duplication
- Changed some methods to reduce complexity with fail fast logic

### CODE REVIEW

**Review by r00ta — APPROVED** [review]

**Review by tteofili — APPROVED** [review]

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/entities/AbstractEntity.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+ * limitations under the License.
+ */
+package org.kie.kogito.explainability.local.counterfactual.entities;
+
+import org.kie.kogito.explainability.model.Feature;
+
+/**
+ * Common class for counterfactual entities
+ */
+public abstract class AbstractEntity<T> implements CounterfactualEntity {
```
Comment (r00ta) [inline explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/entities/AbstractEntity.java:23]: conflicts with https://github.com/kiegroup/kogito-apps/pull/695

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/DummyModelsLimeExplainerTest.java [hunk 2]
```diff
@@ -212,6 +212,7 @@ void testFixedOutput(int seed) throws Exception {
         Prediction prediction = new Prediction(input, outputs.get(0));
         LimeConfig limeConfig = new LimeConfig().withSamples(1000).withPerturbationContext(new PerturbationContext(random, 1));
         LimeExplainer limeExplainer = new LimeExplainer(limeConfig);
+        System.out.println("TIMEOUT " + Config.INSTANCE.getAsyncTimeout() + " " + Config.INSTANCE.getAsyncTimeUnit());
```
Comment (r00ta) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/DummyModelsLimeExplainerTest.java]: Logger?

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/utils/MatrixUtilsTest.java [hunk 3]
```diff
@@ -182,7 +184,9 @@ void testInvertSingular() {
     void testJitterInvert() {
         // since there's some randomness in jitter invert, let's make sure it's stable
         for (int run = 0; run < 100; run++) {
-            double[][] inv = MatrixUtils.jitterInvert(matSquareSingular, 10, 1e-9);
+            Random random = new Random();
+            random.setSeed(run);
+            double[][] inv = MatrixUtils.jitterInvert(matSquareSingular, 10, 1e-9, random);
 
             // since the output of jitterInvert is non-deterministic for singular matrices, check to make sure
```
Comment (r00ta) [inline explainability/explainability-core/src/test/java/org/kie/kogito/explainability/utils/MatrixUtilsTest.java:191]: there are some constants that can be defined as static final in this class

**Review by r00ta — APPROVED** [review]
Thank you!

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/DummyModelsLimeExplainerTest.java (hunk 2 above)
Comment (danielezonca) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/local/lime/DummyModelsLimeExplainerTest.java]: Leftover, thanks, removed 👍

File: explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/entities/AbstractEntity.java (hunk 1 above)
Comment (danielezonca) [reply explainability/explainability-core/src/main/java/org/kie/kogito/explainability/local/counterfactual/entities/AbstractEntity.java:23]: Fine for me to wait 👍

File: explainability/explainability-core/src/test/java/org/kie/kogito/explainability/utils/MatrixUtilsTest.java (hunk 3 above)
Comment (danielezonca) [reply explainability/explainability-core/src/test/java/org/kie/kogito/explainability/utils/MatrixUtilsTest.java:191]: Done

**Review by ruivieira — COMMENTED** [review]
LGTM 👍

**Comment by danielezonca** [comment]
Failed unrelated test `stopWithSolverStarted – org.kie.kogito.taskassigning.service.SolverExecutorTest`

jenkins retest this please

**Pull request merged**